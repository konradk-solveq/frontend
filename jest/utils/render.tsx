import React from 'react';
import {Store} from 'redux';
import {Provider} from 'react-redux';
import {persistStore} from 'redux-persist';
import {PersistGate} from 'redux-persist/integration/react';
import {render, RenderOptions} from '@testing-library/react-native';

import storage, {buildStore, RootState} from '@storage/storage';
import StaticLocationProvider from '@providers/staticLocationProvider/staticLocationProvider';
import TopNotificationProvider from '@providers/topNotificationProvider/TopNotificationProvider';

const persistor = persistStore(storage);

export const renderComponent = async (
    component: any,
    renderOptions?: RenderOptions | undefined,
    initState?: Partial<RootState>,
) => {
    const mockedStore = initState && buildStore(initState);
    const mockerdPersistor = mockedStore && persistStore(mockedStore);

    const wrappedComponent = await render(
        <Provider store={mockedStore || storage}>
            <PersistGate persistor={mockerdPersistor || persistor}>
                <TopNotificationProvider>
                    <StaticLocationProvider>{component}</StaticLocationProvider>
                </TopNotificationProvider>
            </PersistGate>
        </Provider>,
        renderOptions,
    );

    return {
        ...wrappedComponent,
        storage,
    };
};

interface HookWrapperProps {
    children: React.ReactNode;
    initState?: Store<any, any>;
}
export const hookWrapper = ({children, initState}: HookWrapperProps) => (
    <Provider store={initState || storage}>
        <PersistGate persistor={persistor}>{children}</PersistGate>
    </Provider>
);

export const staticLocationProviderWrapper = async (
    component: any,
    renderOptions?: RenderOptions | undefined,
) => {
    const wrappedComponent = await render(
        <Provider store={storage}>
            <PersistGate persistor={persistor}>{component}</PersistGate>
        </Provider>,
        renderOptions,
    );

    return {
        ...wrappedComponent,
        storage,
    };
};

export const rerenderComponent = async (component: any, children: any) => {
    await component.rerender(
        <Provider store={storage}>
            <PersistGate persistor={persistor}>
                <TopNotificationProvider>
                    <StaticLocationProvider>{children}</StaticLocationProvider>
                </TopNotificationProvider>
            </PersistGate>
        </Provider>,
    );
    return;
};

export default renderComponent;
