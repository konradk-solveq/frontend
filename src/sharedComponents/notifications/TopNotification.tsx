import React, {useEffect, useRef, useCallback, useState} from 'react';
import {Platform, StyleSheet, Animated, View, ViewStyle} from 'react-native';

import useStatusBarHeight from '@hooks/statusBarHeight';
import {getFontSize, getHorizontalPx, getVerticalPx} from '@helpers/layoutFoo';

const isIOS = Platform.OS === 'ios';
const expanded = getVerticalPx(60);
const closed = 0;
const defaultDuration = 800;
const defaultHideoutTimeout = 1600;

interface IProps {
    content: string;
    onHideNotification: () => void;
    duration?: number;
    hideTimeout?: number;
    containerStyle?: ViewStyle;
    numberOfLines?: number;
}

const TopNotification: React.FC<IProps> = ({
    content,
    onHideNotification,
    duration,
    hideTimeout,
    containerStyle,
    numberOfLines,
}: IProps) => {
    const animHeightRef = useRef(new Animated.Value(0)).current;
    const animContentOpacitytRef = useRef(new Animated.Value(0)).current;

    const statusbarHeight = useStatusBarHeight();
    const expandedHeight = isIOS ? expanded + statusbarHeight / 2 : expanded;

    const [isVisible, setIsVisible] = useState(false);

    /**
     * Notification is visible when content exits.
     */
    useEffect(() => {
        if (content) {
            setIsVisible(true);
        }
    }, [content]);

    const setNotificationVisibility = useCallback(
        (visibility: boolean) => {
            Animated.timing(animHeightRef, {
                toValue: visibility ? expandedHeight : closed,
                duration: duration || defaultDuration,
                useNativeDriver: false,
            }).start();
            Animated.timing(animContentOpacitytRef, {
                toValue: visibility ? 1 : 0,
                duration: (duration || defaultDuration) / 1.5,
                useNativeDriver: false,
            }).start();
        },
        [animHeightRef, animContentOpacitytRef, duration, expandedHeight],
    );

    useEffect(() => {
        setNotificationVisibility(isVisible);

        /**
         * Auto hide notification
         */
        const hideAfter =
            (hideTimeout || defaultHideoutTimeout) +
            (duration || defaultDuration);
        const t = setTimeout(() => {
            setNotificationVisibility(false);
            setIsVisible(false);
            onHideNotification();
        }, hideAfter);

        return () => {
            clearTimeout(t);
        };
    }, [
        isVisible,
        setNotificationVisibility,
        onHideNotification,
        hideTimeout,
        duration,
    ]);

    return (
        <View style={[styles.container, containerStyle]}>
            <Animated.View
                style={[styles.animContainer, {height: animHeightRef}]}>
                <Animated.Text
                    numberOfLines={numberOfLines || 1}
                    style={[styles.content, {opacity: animContentOpacitytRef}]}>
                    {content}
                </Animated.Text>
            </Animated.View>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        position: 'absolute',
        width: '100%',
    },
    animContainer: {
        position: 'absolute',
        top: 0,
        bottom: 0,
        right: 0,
        left: 0,
        width: '100%',
        height: closed,
        zIndex: 1000,
        backgroundColor: '#313131',
        justifyContent: 'flex-end',
    },
    content: {
        fontFamily: 'DIN2014Narrow-Regular',
        paddingHorizontal: getHorizontalPx(20),
        paddingBottom: getVerticalPx(5),
        fontSize: getFontSize(13),
        letterSpacing: 0.4,
        color: '#ffffff',
        textAlign: 'center',
    },
});

export default React.memo(TopNotification);
