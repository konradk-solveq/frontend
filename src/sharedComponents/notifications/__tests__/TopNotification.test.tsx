import 'react-native';
import React from 'react';
import {cleanup, render} from '@testing-library/react-native';

import TopNotification from '@sharedComponents/notifications/TopNotification';

import {renderHook} from '@testing-library/react-hooks';
import useStatusBarHeight from '@src/hooks/statusBarHeight';
import {hookWrapper} from '@jestUtils/render';

const TEST_MESSAGE = 'Test message';

describe('<TopNotification />', () => {
    describe('Rendering', () => {
        beforeEach(() => {
            const {waitForNextUpdate} = renderHook(() => useStatusBarHeight(), {
                wrapper: hookWrapper,
            });

            waitForNextUpdate();
        });

        it('Should match snapshot', async () => {
            const onHide = jest.fn();

            const component = await render(
                <TopNotification
                    content={TEST_MESSAGE}
                    onHideNotification={onHide}
                />,
            );

            expect(component).toMatchSnapshot();

            const notification = component.getByText('Test message');
            expect(notification).not.toBeNull();
        });

        it('Should not show notification without message content', async () => {
            const onHide = jest.fn();

            const component = await render(
                <TopNotification content={''} onHideNotification={onHide} />,
            );

            expect(component).toMatchSnapshot();

            try {
                component.getByText('Test message');
            } catch (error) {
                expect(error.message).toEqual(
                    'Unable to find an element with text: Test message',
                );
            }
        });
    });

    afterAll(() => {
        cleanup();
    });
});
