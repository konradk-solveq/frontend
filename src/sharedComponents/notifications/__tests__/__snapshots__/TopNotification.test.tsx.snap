// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopNotification /> Rendering Should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "position": "absolute",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "backgroundColor": "#313131",
        "bottom": 0,
        "height": 0,
        "justifyContent": "flex-end",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": "100%",
        "zIndex": 1000,
      }
    }
  >
    <Text
      collapsable={false}
      nativeID="animatedComponent"
      numberOfLines={1}
      style={
        Object {
          "color": "#ffffff",
          "fontFamily": "DIN2014Narrow-Regular",
          "fontSize": 13,
          "letterSpacing": 0.4,
          "opacity": 0,
          "paddingBottom": 7.444196428571429,
          "paddingHorizontal": 36.231884057971016,
          "textAlign": "center",
        }
      }
    >
      Test message
    </Text>
  </View>
</View>
`;

exports[`<TopNotification /> Rendering Should not show notification without message content 1`] = `
<View
  style={
    Array [
      Object {
        "position": "absolute",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "backgroundColor": "#313131",
        "bottom": 0,
        "height": 0,
        "justifyContent": "flex-end",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": "100%",
        "zIndex": 1000,
      }
    }
  >
    <Text
      collapsable={false}
      nativeID="animatedComponent"
      numberOfLines={1}
      style={
        Object {
          "color": "#ffffff",
          "fontFamily": "DIN2014Narrow-Regular",
          "fontSize": 13,
          "letterSpacing": 0.4,
          "opacity": 0,
          "paddingBottom": 7.444196428571429,
          "paddingHorizontal": 36.231884057971016,
          "textAlign": "center",
        }
      }
    >
      
    </Text>
  </View>
</View>
`;
